File: aat-android/src/main/java/ch/bailu/aat/views/preferences/GeneralPreferencesView.java
package ch.bailu.aat.views.preferences;

import android.os.Build;

import ch.bailu.aat.R;
import ch.bailu.aat.activities.ActivityContext;
import ch.bailu.aat.preferences.Storage;
import ch.bailu.aat_lib.preferences.general.SolidWeight;
import ch.bailu.aat.preferences.location.AndroidSolidLocationProvider;
import ch.bailu.aat_lib.preferences.location.SolidAdjustGpsAltitude;
import ch.bailu.aat_lib.preferences.location.SolidAdjustGpsAltitudeValue;
import ch.bailu.aat.preferences.location.SolidGpsTimeFix;
import ch.bailu.aat_lib.preferences.location.SolidPressureAtSeaLevel;
import ch.bailu.aat_lib.preferences.location.SolidProvideAltitude;
import ch.bailu.aat.preferences.system.AndroidSolidDataDirectory;
import ch.bailu.aat_lib.preferences.system.SolidCacheSize;
import ch.bailu.aat.preferences.system.SolidExternalDirectory;
import ch.bailu.aat_lib.preferences.system.SolidStatusMessages;
import ch.bailu.aat.util.ui.UiTheme;
import ch.bailu.aat.views.SensorListView;
import ch.bailu.aat_lib.gpx.InfoID;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.preferences.general.SolidPresetCount;
import ch.bailu.aat_lib.preferences.general.SolidUnit;


public class GeneralPreferencesView extends VerticalScrollView {

    public GeneralPreferencesView(ActivityContext acontext, UiTheme theme) {
        super(acontext);

        StorageInterface storage = new Storage(acontext);

        add(new TitleView(acontext, R.string.p_general, theme));
        add(new SolidIndexListView(acontext,new SolidUnit(storage), theme));
        add(new SolidTextInputView(acontext,new SolidWeight(storage),
                SolidTextInputDialog.INTEGER, theme));

        add(new SolidIndexListView(acontext,new SolidPresetCount(storage), theme));
        add(new SolidIndexListView(acontext,new SolidStatusMessages(storage), theme));

        add(new TitleView(acontext, "GPS", theme));
        add(new SolidIndexListView(acontext,new AndroidSolidLocationProvider(acontext), theme));
        add(new SolidCheckBox(acontext,new SolidGpsTimeFix(storage), theme));
        add(new SolidCheckBox(acontext,new SolidAdjustGpsAltitude(storage), theme));
        add(new SolidTextInputView(acontext,
                new SolidAdjustGpsAltitudeValue(storage, new SolidUnit(storage).getIndex()),
                SolidTextInputDialog.INTEGER_SIGNED,
                theme));

        add(new SolidTextInputView(acontext,new SolidPressureAtSeaLevel(storage),
                SolidTextInputDialog.FLOAT,
                theme));

        add(new SolidTextInputView(acontext,new SolidProvideAltitude(storage, SolidUnit.SI),
                SolidTextInputDialog.INTEGER_SIGNED,
                theme));
        add(new SolidTextInputView(acontext,new SolidProvideAltitude(storage, SolidUnit.IMPERIAL),
                SolidTextInputDialog.INTEGER_SIGNED,
                theme));

        if (Build.VERSION.SDK_INT >= 18) {
            add(new TitleView(acontext, R.string.sensors, theme));
            ScanBluetoothView scan = new ScanBluetoothView(
                    acontext.getServiceContext(), theme
            );
            ConnectToSensorsView updateConnection = new ConnectToSensorsView(
                    acontext.getServiceContext(),
                    theme
            );
            SensorListView sensors = new SensorListView(acontext.getServiceContext(), theme);

            add(scan);
            add(updateConnection);
            add(sensors);
            acontext.addTarget(scan, InfoID.SENSORS);
            acontext.addTarget(updateConnection, InfoID.SENSORS);
            acontext.addTarget(sensors, InfoID.SENSORS);
        }

        add(new TitleView(acontext, R.string.files, theme));
        add(new SolidDirectoryViewSAF(acontext, new AndroidSolidDataDirectory(acontext), theme));
        add(new SolidDirectoryViewSAF(acontext, new SolidExternalDirectory(acontext), theme));

        add(new SolidIndexListView(acontext, new SolidCacheSize(storage), theme));
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/views/preferences/GeneralPreferencesView.kt
package ch.bailu.aat.views.preferences

import ch.bailu.aat.R
import ch.bailu.aat.activities.ActivityContext
import ch.bailu.aat.preferences.Storage
import ch.bailu.aat.preferences.location.AndroidSolidLocationProvider
import ch.bailu.aat.preferences.location.SolidGpsTimeFix
import ch.bailu.aat.preferences.system.AndroidSolidDataDirectoryDefault
import ch.bailu.aat.preferences.system.SolidExternalDirectory
import ch.bailu.aat.util.ui.theme.UiTheme
import ch.bailu.aat.views.list.SensorListView
import ch.bailu.aat.views.preferences.dialog.SolidTextInputDialog
import ch.bailu.aat_lib.gpx.information.InfoID
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.preferences.general.SolidPresetCount
import ch.bailu.aat_lib.preferences.general.SolidUnit
import ch.bailu.aat_lib.preferences.general.SolidWeight
import ch.bailu.aat_lib.preferences.location.SolidAdjustGpsAltitude
import ch.bailu.aat_lib.preferences.location.SolidAdjustGpsAltitudeValue
import ch.bailu.aat_lib.preferences.location.SolidPressureAtSeaLevel
import ch.bailu.aat_lib.preferences.location.SolidProvideAltitude
import ch.bailu.aat_lib.preferences.system.SolidCacheSize
import ch.bailu.aat_lib.preferences.system.SolidDataDirectory
import ch.bailu.aat_lib.preferences.system.SolidStatusMessages
import ch.bailu.aat_lib.resources.ToDo
import ch.bailu.foc_android.FocAndroidFactory

class GeneralPreferencesView(acontext: ActivityContext, theme: UiTheme) :
    VerticalScrollView(acontext) {
    init {
        val storage: StorageInterface = Storage(acontext)
        add(TitleView(acontext, R.string.p_general, theme))
        add(SolidIndexListView(acontext, SolidUnit(storage), theme))
        add(
            SolidTextInputView(
                acontext, SolidWeight(storage),
                SolidTextInputDialog.INTEGER, theme
            )
        )
        add(SolidIndexListView(acontext, SolidPresetCount(storage), theme))
        add(SolidIndexListView(acontext, SolidStatusMessages(storage), theme))
        add(TitleView(acontext, "GPS", theme))
        add(SolidIndexListView(acontext, AndroidSolidLocationProvider(acontext), theme))
        add(SolidCheckBox(acontext, SolidGpsTimeFix(storage), theme))
        add(SolidCheckBox(acontext, SolidAdjustGpsAltitude(storage), theme))
        add(
            SolidTextInputView(
                acontext,
                SolidAdjustGpsAltitudeValue(storage, SolidUnit(storage).index),
                SolidTextInputDialog.INTEGER_SIGNED,
                theme
            )
        )
        add(
            SolidTextInputView(
                acontext, SolidPressureAtSeaLevel(storage),
                SolidTextInputDialog.FLOAT,
                theme
            )
        )
        add(
            SolidTextInputView(
                acontext, SolidProvideAltitude(storage, SolidUnit.SI),
                SolidTextInputDialog.INTEGER_SIGNED,
                theme
            )
        )
        add(
            SolidTextInputView(
                acontext, SolidProvideAltitude(storage, SolidUnit.IMPERIAL),
                SolidTextInputDialog.INTEGER_SIGNED,
                theme
            )
        )
        add(TitleView(acontext, R.string.sensors, theme))
        val scan = ScanBluetoothView(
            acontext.serviceContext, theme
        )
        val updateConnection = ConnectToSensorsView(
            acontext.serviceContext,
            theme
        )
        val sensors = SensorListView(
            acontext.serviceContext,
            theme
        )
        add(scan)
        add(updateConnection)
        add(sensors)
        acontext.dispatcher.addTarget(scan, InfoID.SENSORS)
        acontext.dispatcher.addTarget(updateConnection, InfoID.SENSORS)
        acontext.dispatcher.addTarget(sensors, InfoID.SENSORS)

        add(TitleView(acontext, ToDo.translate("App permissions"), theme))
        add(CheckPermissionsView(acontext.appContext,acontext, theme))
        add(RequestPermissionsView(acontext, theme))

        add(TitleView(acontext, R.string.files, theme))
        add(SolidDirectoryViewSAF(acontext, SolidDataDirectory(AndroidSolidDataDirectoryDefault(context), FocAndroidFactory(context)), theme))
        add(SolidDirectoryViewSAF(acontext, SolidExternalDirectory(acontext), theme))
        add(SolidIndexListView(acontext, SolidCacheSize(storage), theme))
    }
}
